python/pyarrow/tests/test_flight.py

```diff
@@ -471,3 +516,26 @@ def test_location_invalid():
     server = ConstantFlightServer()
     with pytest.raises(pa.ArrowInvalid, match=".*Cannot parse URI:.*"):
         server.run("%")
+
+
+@pytest.mark.skipif(
+    test_resource_certs() is None,
+    reason="Test resources are unavailable; set ARROW_TEST_DATA.",
+)
+def test_tls():
+    """Try a simple do_get call over TLS."""
+    table = simple_ints_table()
+    certs = test_resource_certs()
+
+    with flight_server(
+            ConstantFlightServer, tls_certificates=certs["certificates"]
+    ) as server_location:
+        options = flight.FlightCallOptions(timeout=0.5)
+        client = flight.FlightClient.connect(server_location)
+        with pytest.raises(pa.ArrowIOError, match="Deadline Exceeded"):
```

Comment:
Is there a point in testing per-call timeout here?

Reply:
I added an explanatory comment, but the purpose of the timeout is to make sure gRPC doesn't block there for a long time trying to reconnect over and over. (Not sure why it doesn't fail-fast in these cases.)

Reply:
Hmm, looks like we set gRPC's [`wait_for_ready` option](https://grpc.github.io/grpc/cpp/classgrpc_1_1_client_context.html#aa342af7349c88727e8bd4c765a000ae5), which is why it doesn't fail-fast. But unsetting that causes C++ tests to fail due to a race condition between starting the server subprocess and creating the client. I'll convert the tests to not use the subprocess (should make them run faster too), then drop the option and the timeout here.

Reply:
Not using a subprocess is nice, but we'll still need to allow retries, otherwise the race condition will still be there AFAIK.

Reply:
I'm going to ping the server (with a short timeout) until a request succeeds, rather than rely on experimental gRPC options that we don't have much control over (again, in my opinion, blocking for some unbounded amount of time is a rather pathological way to handle connection failure...it also leads to more cases where Ctrl-C doesn't work in Python)

Reply:
Actually as implemented, creating a server immediately starts it, so there's no need for even that...

Reply:
It starts the thread, but is the listening socket already created before?

Reply:
Yes - `FlightServerBase::Init` calls `BuildAndStart`, so all `Serve` *actually* does is set up signals and block until the server shuts down. (maybe we shouldn't rely on that...)